File: js/day80.js
// Ask for a password input, if new, require strong password and register, if password exists(hashed password txt file), require existing password

const readline = require('readline');
const fs = require('fs');
const crypto = require('crypto');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const FILE_PATH = './pass.txt';

// A simple function to turn a password into a hashed string
function simpleHash(password) {
    return crypto.createHash('sha256').update(password).digest('hex');
}

// Ask for the username first
rl.question('Enter Username: ', (username) => {
    username = username.trim();

    // Check if our password file exists, then read it
    let fileData = '';
    if (fs.existsSync(FILE_PATH)) {
        fileData = fs.readFileSync(FILE_PATH, 'utf8');
    }

    // Search the file to see if the user already exists
    let savedHash = '';
    const lines = fileData.split('\n');
    for (let line of lines) {
        const parts = line.split(':');
        if (parts[0] === username) {
            savedHash = parts[1]; // Found the existing hashed password
            break;
        }
    }

    if (savedHash) {
        // --- LOGIN FLOW ---
        console.log('User found! Please login.');
        rl.question('Enter Password: ', (password) => {
            
            // Hash the input and compare it to what is in the file
            if (simpleHash(password) === savedHash) {
                console.log('Success: Logged in successfully!');
            } else {
                console.log('Error: Incorrect password.');
            }
            rl.close(); // Close the input tool
        });

    } else {
        // --- REGISTRATION FLOW ---
        console.log('New user detected! Please register.');
        console.log('Rule: Password must be at least 8 characters long, have a number, and a symbol.');
        
        rl.question('Create Strong Password: ', (password) => {
            // Simple validation check: 8+ chars, contains a number, contains a special character
            const hasNumber = /[0-9]/.test(password);
            const hasSymbol = /[^A-Za-z0-9]/.test(password);

            if (password.length >= 8 && hasNumber && hasSymbol) {
                const newHash = simpleHash(password);
                
                // Save the user and hash to the file
                fs.appendFileSync(FILE_PATH, username + ':' + newHash + '\n');
                console.log('Success: Account created and saved to pass.txt!');
            } else {
                console.log('Error: Password is too weak. Registration cancelled.');
            }
            rl.close();
        });
    }
});

File: python/day80.py
import hashlib
import os
import re

# Ask for a password input, if new, require strong password and register, if password exists(hashed password txt file), require existing password

FILE_PATH = "pass.txt"

def simple_hash(password):
    """Turns a password string into a secure SHA-256 hex string."""
    return hashlib.sha256(password.encode()).hexdigest()

def main():
    print("=== Terminal Authentication System ===")
    username = input("Enter Username: ").strip()
    
    if not username:
        print("Username cannot be empty.")
        return

    # 1. Read existing users from the file
    users = {}
    if os.path.exists(FILE_PATH):
        with open(FILE_PATH, "r", encoding="utf-8") as file:
            for line in file:
                line = line.strip()
                if ":" in line:
                    parts = line.split(":")
                    users[parts[0]] = parts[1] # format -> username: hashed_password

    # 2. Routing based on whether the user exists
    if username in users:
        # --- LOGIN FLOW ---
        print("User found! Please login.")
        password = input("Enter Password: ")
        
        if simple_hash(password) == users[username]:
            print("\033[92mSuccess: Logged in successfully!\033[0m")
        else:
            print("\033[91mError: Incorrect password.\033[0m")
            
    else:
        # --- REGISTRATION FLOW ---
        print("New user detected! Please register.")
        print("Rule: Min 8 chars, 1 uppercase, 1 lowercase, 1 number, 1 symbol.")
        
        password = input("Create Strong Password: ")
        
        # Strict validation check
        has_upper = any(c.isupper() for c in password)
        has_lower = any(c.islower() for c in password)
        has_digit = any(c.isdigit() for c in password)
        has_symbol = bool(re.search(r"[^A-Za-z0-9]", password))
        
        if len(password) >= 8 and has_upper and has_lower and has_digit and has_symbol:
            new_hash = simple_hash(password)
            
            # Append user to file securely
            with open(FILE_PATH, "a", encoding="utf-8") as file:
                file.write(f"{username}:{new_hash}\n")
                
            print("\033[92mSuccess: Account created and saved to pass.txt!\033[0m")
        else:
            print("\033[91mError: Password does not meet requirements. Registration cancelled.\033[0m")

if __name__ == "__main__":
    main()
